earth2grid/healpix/_padding/pure_python.py

```diff
@@ -0,0 +1,155 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import torch
+
+from earth2grid.healpix import core
+
+# most routines in this module use the standard origin=S convention
+PIXEL_ORDER = core.XY()
+
+
+def _xy_with_filled_tile(nside, x1, y1, f1):
+    """Handles an points with missing tile information following the HPXPAD strategy
+
+    Missing tiles are defined for face >= 12. 12-16 are the N missing tiles, and
+    16-20 the south missing tiles (from W to east).
+
+    Since there is an ambiguity return both x and y.
+    """
+
+    # handle missing tiles
+    # for N tiles
+    # f(x, y) is filled by shuffling from the left
+    # case x > y: (x, y) -> (y, )
+    # examples  (for nside = 4)
+    #   (3, 1)-> (0, 1)
+    #   (3, 2) -> (1, 2)
+    #   (3, 3) -> (2, 3)
+    # generalize
+    #   (i, j)-> (i + j, j)  in the missing face
+    #   (i' - j, j) -> (i', j)
+
+    is_missing_n_pole_tile = (f1 >= 12) & (f1 < 16)
+    west_face = torch.where(is_missing_n_pole_tile, f1 - 13, 0) % 4
+    east_face = (west_face + 1) % 4
+
+    # two sets of indices
+    def _pad_from_west(x1, y1, west_face):
+        f_west = torch.where(is_missing_n_pole_tile & (x1 <= y1), west_face, f1)
+        x_west = torch.where(is_missing_n_pole_tile & (x1 < y1), (x1 - y1) % nside, x1)
+        x_west = torch.where(is_missing_n_pole_tile & (x1 == y1), nside - 1, x_west)
+        y_west = y1
+        return x_west, y_west, f_west
+
+    x_west, y_west, f_west = _pad_from_west(x1, y1, west_face)
+    y_east, x_east, f_east = _pad_from_west(y1, x1, east_face)
+
+    # S pole
+    is_missing_s_pole_tile = (f1 >= 16) & (f1 < 20)
+    east_face = (f1 - 16) % 4 + 8
+    west_face = (east_face - 9) % 4 + 8
+
+    # two sets of indices
+    def _pad_from_east(x1, y1, east_face, f1):
+        """Test cases
+
+        (1, 0) -> (0, 0)
+        (3, 2) -> (0, 2)
+        """
+        f_west = torch.where(is_missing_s_pole_tile & (x1 >= y1), east_face, f1)
+        # x_west = torch.where(is_missing_s_pole_tile & (x1 > y1), 1(x1-y1) %nside, x1)
+        x_west = torch.where(is_missing_s_pole_tile & (x1 > y1), (x1 - y1 - 1) % nside, x1)
+        x_west = torch.where(is_missing_s_pole_tile & (x1 == y1), 0, x_west)
+        y_west = y1
+        return x_west, y_west, f_west
+
+    x_west, y_west, f_west = _pad_from_east(x_west, y_west, east_face, f_west)
```

Comment:
Is this correct? (from_east vs from_west)

Reply:
confusing names, but it is correct.